scraper/company_scraper_test.go

```diff
@@ -0,0 +1,334 @@
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"testing"
+
+	"github.com/Businge931/company-email-scraper/config"
+	"github.com/google/go-querystring/query"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestReadCompanyNames(t *testing.T) {
+	tests := map[string]struct {
+		companyNames []string
+		expectError  bool
+	}{
+		"ValidCompanyNames": {
+			companyNames: []string{"stanbic bank", "Pearl Technologies", "clinicpesa"},
+			expectError:  false,
+		},
+		"EmptyFile": {
+			companyNames: []string{},
+			expectError:  false,
+		},
+		"FileNotFound": {
+			expectError: true,
+		},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			// Prepare a temporary directory and file for testing, except for the "FileNotFound" case
+			var tempFile string
+			if name != "FileNotFound" {
+				tempDir := "companies-list"
+				tempFile = tempDir + "/input.txt"
+
+				if err := os.MkdirAll(tempDir, os.ModePerm); err != nil {
+					t.Fatalf("Failed to create temp directory: %v", err)
+				}
+				defer os.RemoveAll(tempDir) // Clean up the temp directory after the test
+
+				file, err := os.Create(tempFile)
+				if err != nil {
+					t.Fatalf("Failed to create temp file: %v", err)
+				}
+				defer file.Close()
+
+				// Write the company names to the file
+				for _, name := range tt.companyNames {
+					if _, err := file.WriteString(name + "\n"); err != nil {
+						t.Fatalf("Failed to write to temp file: %v", err)
+					}
+				}
+			} else {
+				// Set a non-existing file path for "FileNotFound" case
+				tempFile = "non-existing-file.txt"
+			}
+
+			// Call the function under test
+			actualNames, err := ReadCompanyNames(tempFile)
+
+			// Check for expected error
+			if tt.expectError {
+				if err == nil {
+					t.Fatalf("Expected an error, but got nil")
+				}
+				return
+			} else {
+				if err != nil {
+					t.Fatalf("Did not expect an error, but got: %v", err)
+				}
+			}
+
+			// Check the length and content of slices
+			if len(actualNames) != len(tt.companyNames) {
+				t.Errorf("Expected %d company names, but got %d", len(tt.companyNames), len(actualNames))
+			}
+
+			for i, name := range tt.companyNames {
+				if actualNames[i] != name {
+					t.Errorf("Expected company name %q at index %d, but got %q", name, i, actualNames[i])
+				}
+			}
+		})
+	}
+}
+
+func TestGetSearchResults(t *testing.T) {
+	tests := map[string]struct {
+		companyName   string
+		apiKey        string
+		mockServer    *httptest.Server
+		expectedError string
+		expectedURL   string
+	}{
+		"Successful search with valid API key and company name": {
+			companyName: "testCompany",
+			apiKey:      "validApiKey",
+			mockServer: httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusOK)
+				json.NewEncoder(w).Encode(SerpAPIResponse{
+					Organic: []struct {
+						Link string `json:"link"`
+					}{
+						{Link: "https://example.com"},
+					},
+				})
+			})),
```

Comment:
We will generally optimize this through this approach https://www.thegreatcodeadventure.com/mocking-http-requests-in-golang/ which will require introduction of an interface. Will discuss this as well